src/monodroid/jni/monodroid-glue.cc

```diff
@@ -1219,145 +1202,176 @@ MonodroidRuntime::set_environment_variable_for_directory (const char *name, jstr
+	size_t home_len = strlen (homeDir.get_cstr ());
+
+	constexpr char XDG_DATA_HOME[] = "XDG_DATA_HOME";
+	constexpr char HOME_PATH[] = ".local/share";
+	constexpr size_t HOME_PATH_LEN = sizeof(HOME_PATH) - 1;
+	create_xdg_directory (homeDir, home_len, HOME_PATH, HOME_PATH_LEN, XDG_DATA_HOME);
+
+	constexpr char XDG_CONFIG_HOME[] = "XDG_CONFIG_HOME";
+	constexpr char CONFIG_PATH[] = ".config";
+	constexpr size_t CONFIG_PATH_LEN = sizeof(CONFIG_PATH) - 1;
+	create_xdg_directory (homeDir, home_len, CONFIG_PATH, CONFIG_PATH_LEN, XDG_CONFIG_HOME);
 }
 
 #if DEBUG
 void
 MonodroidRuntime::set_debug_env_vars (void)
 {
-	char *value;
-
-	if (androidSystem.monodroid_get_system_property (Debug::DEBUG_MONO_ENV_PROPERTY, &value) == 0)
+	dynamic_local_string<PROPERTY_VALUE_BUFFER_LEN> value;
+	if (androidSystem.monodroid_get_system_property (Debug::DEBUG_MONO_ENV_PROPERTY, value) == 0)
 		return;
 
-	char **args = utils.monodroid_strsplit (value, "|", 0);
-	delete[] value;
+	auto log_envvar = [](const char *name, const char *v) {
+		log_info (LOG_DEFAULT, "Env variable '%s' set to '%s'.", name, v);
+	};
+
+	string_segment arg_token;
+	while (value.next_token ('|', arg_token)) {
+		static_local_string<SMALL_STRING_PARSE_BUFFER_LEN> arg (arg_token.length ());
+		arg.assign (arg_token.start (), arg_token.length ());
 
-	for (char **ptr = args; ptr && *ptr; ptr++) {
-		char *arg = *ptr;
-		char *v = strchr (arg, '=');
-		if (v) {
-			*v = '\0';
-			++v;
-		} else {
+		ssize_t idx = arg.index_of ('=');
+		size_t index = static_cast<size_t>(idx);
+		if (idx < 0 || index == arg.length () - 1) {
+			// ’name’ or ’name=’
 			constexpr char one[] = "1";
-			v = const_cast<char*> (one);
+			if (index == arg.length () - 1) {
+				arg[index] = '\0';
+			}
+			setenv (arg.get (), one, 1);
+			log_envvar (arg.get (), one);
+		} else if (index == 0) {
+			// ’=value’
+			log_warn (LOG_DEFAULT, "Attempt to set environment variable without specifying name: '%s'", arg.get ());
+		} else {
+			// ’name=value’
+			arg[index] = '\0';
+			const char *v = arg.get () + idx + 1;
+			setenv (arg.get (), v, 1);
+			log_envvar (arg.get (), v);
 		}
-		setenv (arg, v, 1);
-		log_info (LOG_DEFAULT, "Env variable '%s' set to '%s'.", arg, v);
 	}
-	utils.monodroid_strfreev (args);
 }
 #endif /* DEBUG */
 
 inline void
 MonodroidRuntime::set_trace_options (void)
 {
-	char *value;
-
-	if (androidSystem.monodroid_get_system_property (Debug::DEBUG_MONO_TRACE_PROPERTY, &value) == 0)
+	dynamic_local_string<PROPERTY_VALUE_BUFFER_LEN> value;
+	if (androidSystem.monodroid_get_system_property (Debug::DEBUG_MONO_TRACE_PROPERTY, value) == 0)
 		return;
 
-	mono_jit_set_trace_options (value);
-	delete[] value;
+	mono_jit_set_trace_options (value.get ());
 }
 
 inline void
 MonodroidRuntime::set_profile_options ()
 {
-	constexpr const char mlpd_ext[] = "mlpd";
-	constexpr const char aot_ext[] = "aotprofile";
+	log_warn (LOG_DEFAULT, __PRETTY_FUNCTION__);
```

Comment:
I assume we shouldn't preserve this statement, in which case I would suggest using `log_warn(LOG_DEFAULT, "#grendel"`-prefixed messages, so that it's easier to "see" the log messages which should be *removed* before merging.